stdlib/dynarray.ml

```diff
@@ -1247,3 +1264,14 @@ let to_seq_rev_reentrant a =
     end
   in
   aux (length a - 1)
+
+external unsafe_iarray_of_array : 'a array -> 'a iarray = "%opaque"
+
+let unsafe_to_iarray ~capacity (f : 'a t -> unit) =
+  let a = create () in
+  set_capacity a capacity;
+  f a;
+  fit_capacity a;
+  let Pack ({arr; length = _; dummy = _} as a_) = a in
+  a_.length <- 0;
+  unsafe_iarray_of_array (Dummy.Array.unsafe_to_array_without_dummies arr)
```

Comment:
You should steal the backing array from the dynarray, which can be done by calling `reset a` instead of `a_.length <- 0`. If you do this, then there is no problem with `f` leaking `a`, the leaked aliases will start operating on an independent backing array.

Reply:
So here would be a suggested implementation (untested):

```ocaml
let unsafe_to_iarray ~capacity (f : 'a t -> unit) =
  let a = create () in
  set_capacity a capacity;
  f a;
  let Pack {arr; length; dummy} = a in
  reset a;
  (* at this point further updates to [a] will not mutate [arr]. *)
  check_valid_length length arr;
  let arr_values : ('a, _) with_dummy =
    if Array.length arr = length then arr
    else Dummy.Array.prefix arr length
  in
  let values : 'a array =
    Dummy.Array.unsafe_nocopy_to_array arr_values in
  unsafe_iarray_of_array values
```

Reply:
(Note: I edited the version above slightly after the fact, so maybe don't rely on what is in your email reader.

Reply:
Using `reset` to replace the backing array is a good idea, thanks.